modules/tinymce/src/core/main/ts/selection/MultiClickSelection.ts

```diff
@@ -1,44 +1,78 @@
-import { SugarElement } from '@ephox/sugar';
+import { Compare, PredicateFind, SugarElement } from '@ephox/sugar';
 
+import DomTreeWalker from '../api/dom/TreeWalker';
 import Editor from '../api/Editor';
-import * as CaretFinder from '../caret/CaretFinder';
+import { isCaretCandidate } from '../caret/CaretCandidate';
 import { CaretPosition } from '../caret/CaretPosition';
-import { isTextBlock } from '../dom/ElementType';
+import { isBlock } from '../dom/ElementType';
 import * as NodeType from '../dom/NodeType';
 import * as RangeNormalizer from './RangeNormalizer';
 
-const isTextBlockNode = (node: Node): node is Element => NodeType.isElement(node) && isTextBlock(SugarElement.fromDom(node));
+const isBr = NodeType.isBr;
+const isText = NodeType.isText;
+const isContentEditableFalse = (elm: SugarElement<Node>): boolean => NodeType.isContentEditableFalse(elm.dom);
+const isRoot = (rootNode: Node) => (elm: SugarElement<Node>): boolean => Compare.eq(SugarElement.fromDom(rootNode), elm);
 
-const normalizeSelection = (editor: Editor) => {
-  const rng = editor.selection.getRng();
-  const startPos = CaretPosition.fromRangeStart(rng);
-  const endPos = CaretPosition.fromRangeEnd(rng);
+const getClosestBlock = (node: Node, rootNode: Node): Node =>
+  PredicateFind.closest(SugarElement.fromDom(node), isBlock, isRoot(rootNode))
+    .getOr(SugarElement.fromDom(rootNode)).dom;
 
-  if (CaretPosition.isElementPosition(startPos)) {
-    const container = startPos.container();
-    if (isTextBlockNode(container)) {
-      CaretFinder.firstPositionIn(container).each((pos) => rng.setStart(pos.container(), pos.offset()));
+const getClosestCef = (node: Node, rootNode: Node) =>
+  PredicateFind.closest(SugarElement.fromDom(node), isContentEditableFalse, isRoot(rootNode));
+
+const findEdgeCaretCandidate = (startNode: Node, scope: Node, forward: boolean): Node => {
+  const walker = new DomTreeWalker(startNode, scope);
+  const next = forward ? walker.next.bind(walker) : walker.prev.bind(walker);
+  let result: Node = startNode;
+  for (let current = forward ? startNode : next(); current && !isBr(current); current = next()) {
+    if (isCaretCandidate(current)) {
+      result = current;
     }
   }
+  return result;
+};
 
-  if (CaretPosition.isElementPosition(endPos)) {
-    const container = startPos.container();
-    if (isTextBlockNode(container)) {
-      CaretFinder.lastPositionIn(container).each((pos) => rng.setEnd(pos.container(), pos.offset()));
+const findClosestBlockRange = (startRng: Range, rootNode: Node) => {
+  const startPos = CaretPosition.fromRangeStart(startRng);
+  const clickNode = startPos.getNode();
+  const scope = getClosestBlock(clickNode, rootNode);
+  const startNode = findEdgeCaretCandidate(clickNode, scope, false);
+  const endNode = findEdgeCaretCandidate(clickNode, scope, true);
+
+  const rng = document.createRange();
+  getClosestCef(startNode, scope).fold(() => {
```

Comment:
Maybe extremely picky but I tend to format each case of the fold like this:
```
getClosestCef(startNode, scope).fold(
  () => {
    if (isText(startNode)) {
      rng.setStart(startNode, 0);
    } else {
      rng.setStartBefore(startNode);
    }
  },
  (cef) => rng.setStartBefore(cef.dom)
);
```
But maybe that is a old style I used when we had adts faked using folds like this where each case would be it's own line.